src/realm/object-store/sectioned_results.hpp

```diff
@@ -0,0 +1,183 @@
+//
+// Copyright 2022 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_SECTIONED_RESULTS_HPP
+#define REALM_SECTIONED_RESULTS_HPP
+
+#include <realm/util/functional.hpp>
+
+namespace realm {
+class Mixed;
+class Results;
+class SectionedResults;
+struct Section;
+struct SectionedResultsChangeSet;
+
+using SectionedResultsNotificatonCallback = util::UniqueFunction<void(SectionedResultsChangeSet, std::exception_ptr)>;
+
+/**
+ * An instance of ResultsSection gives access to elements in the underlying collection that belong to a given section.
+ *
+ * A ResultsSection is only valid as long as it's `SectionedResults` parent stays alive.
+ *
+ * You can register a notification callback for changes to elements in a specific `ResultsSection`.
+ */
+class ResultsSection {
+public:
+    ResultsSection() = default;
+
+    /// Retrieve an element from the section for a given index.
+    Mixed operator[](size_t idx) const;
+
+    /// The key identifying this section.
+    Mixed key();
+
+    /// The total count of elements in this section.
+    size_t size();
+
+    /**
+     * Create an async query from this ResultsSection
+     * The query will be run on a background thread and delivered to the callback,
+     * and then rerun after each commit (if needed) and redelivered if it changed
+     *
+     * @param callback The function to execute when a insertions, modification or deletion in this `ResultsSection`
+     * was detected.
+     * @param key_path_array A filter that can be applied to make sure the `SectionedResultsNotificatonCallback` is
+     * only executed when the property in the filter is changed but not otherwise.
+     *
+     * @return A `NotificationToken` that is used to identify this callback. This token can be used to remove the
+     * callback via `remove_callback`.
+     */
+    NotificationToken add_notification_callback(SectionedResultsNotificatonCallback callback,
+                                                KeyPathArray key_path_array = {}) &;
+
+    realm::ThreadSafeReference thread_safe_reference();
+    bool is_valid() const;
+
+private:
+    friend class SectionedResults;
+    ResultsSection(SectionedResults* parent, size_t index)
+        : m_parent(parent)
+        , m_index(index)
+    {
+    }
+
+    SectionedResults* m_parent;
+    size_t m_index;
```

Comment:
Is binding a ResultsSection to a section index rather than a section key actually the desired behavior? It might be, but it also seems a little weird. The docs should probably explicitly call it out, especially with regards to how notifications on a section behave.

Reply:
I think I see what you're saying. A section can shift indexes and still be perfectly valid, and binding it to an index will create invalid behaviour with notifications, I'll change it to use the key instead.